---
title: "**Gene-culture** coevolution (GC-coev)"
author: "Thomas Huet"
bibliography:
- 'refs.bib'
output: 
  bookdown::html_document2:
    number_sections: false
    keep_md: true
  # html_document:
    toc: true
    toc_float: 
      collapsed: false
      smooth_scroll: false
---

<style>
.html-widget {
margin: auto;
}
.leaflet .legend {
text-align:left;
line-height: 12px;
font-size: 12px;
}
</style>

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(leaflet)
library(shiny)
library(kableExtra)
library(dplyr)
library(knitr)
library(tidyverse)
library(phylotools)
library(igraph)
library(RColorBrewer)
library(Bchron)
library(rcarbon)

metadata <- "mtdb_metadata.csv"

blank_theme <- theme_minimal()+
  theme(
    axis.title.x = element_blank(),
    axis.title.y = element_blank(),
    panel.border = element_blank(),
    panel.grid=element_blank(),
    axis.ticks = element_blank(),
    plot.title=element_text(size=14, face="bold")
  )
```

Gene-culture coevolution (GC-coev) is one of the main challenge of evolutionary archaeology. It requires the cross study of aDNA ('***Who ?***') and the different aspects of the culture (['***What ?***'](https://github.com/zoometh/thomashuet/blob/main/README.md#what)). 

This doc is a HTML presentation host on [GitHub](https://github.com/zoometh/aDNA), which brings together R + Leaflet coding showing how to download data from the Ancient mtDNA database (AmtDB) and we conduct and overview of GC-coev issues concerning the transition between hunter-gatherers (HG) and early farmers (EF) in the Central and West Europe.

# The AmtDB dataset

The mitochondrial DNA (mtDNA) makes it possible to trace the maternal line. It passes from the mother to her children (of both sexes). Published mitochondrial sequences coming from the ancient DNA samples (aDNA) can be download from the [Ancient mtDNA database (AmtDB)](https://amtdb.org/records/). This database gathers . The whole dataset is composed by the [data](#mt.data) and the [metadata](#mt.meta)


## data {#mt.data}

The data file [mtdb_1511.fasta](mtdb_1511.fasta) is downloaded from the AmtDB. The current metadata format is .fasta, a text-based format for representing either nucleotide sequences or peptide sequences, in which base pairs or amino acids are represented using single-letter codes. FASTA formats can be read with the [phylotools](https://github.com/helixcn/phylotools) package. A sequence in FASTA format begins with a single-line description, followed by lines of sequence data:

```{r data1, echo=T}
fasta <- phylotools::read.fasta("mtdb_1511.fasta")
fasta$seq.name[1]
substr(fasta$seq.text[1], 1, 250)
```
Each of these identifiers is a unique key for a sample. The data file counts <span style='color:grey'>`r length(fasta$seq.text)`</span> different identifiers. These identifiers which allow to join the second dataset, [metadata](#mt.meta). We will use this latter for cross-analysis

## metadata {#mt.meta}

The metadata file [metadata](mtdb_metadata.csv) is downloaded from the AmtDB. The current metadata format is .csv

```{r meta1, echo=F, fig.align='center', fig.cap="Preview of the current [NeoNet dataset](#panel.data) (work in progress) using RShiny and Leaflet in R"}
df <- read.csv(metadata, encoding = "UTF-8")
kable(df[sample(nrow(df), 3), ],"html",
      row.names = F,
      caption = "aDNA metadata sample") %>%
  kable_styling(full_width = FALSE, position = "center", font_size=11)
```

It counts <span style='color:grey'>`r nrow(df)`</span> samples and <span style='color:grey'>`r ncol(df)`</span> columns. 

# mtDNA *and* culture tag {#cross.mt.DNA.culture}

Within the AmtDB database, one (1) field concerns the mtDNA hg ('mt_hg') and two (2) other fields give insights on the cultural membership of the sample:

1. 'culture' counts <span style='color:grey'>`r length(unique(df$culture))`</span> distinct values
2. 'epoch' is more generic and counts <span style='color:grey'>`r length(unique(df$epoch))`</span> distinct values 

To illustrate GC-coev, we will focus on the culture tags (i.e., 'culture' column) of the Mesolithic/Neolithic transition (i.e., 'epoch' == 'Mesolithic' or 'Neolithic'). Our region of interest is the Central and Western Mediterranean (i.e., from Greece to Spain). We will remove sample where the 'mt_hg' is empty, and to have an easily legible dataset we will also remove *unicum* (Freq == 1) in the 'culture' field

```{r def, echo=T}
df.MesoNeo <- df %>% 
  filter(mt_hg != "" & !is.na(mt_hg)) %>%
  filter(epoch == 'Mesolithic' | epoch == 'Neolithic') %>%
  filter(latitude < 46 & latitude > 35) %>%
  filter(longitude > -6 & longitude < 28)
df.MesoNeo <- df.MesoNeo %>%  
  group_by(culture) %>%
  filter(n() > 1)
MNcultures <- as.data.frame(table(df.MesoNeo$culture),
                            stringsAsFactors = F)
```

After processing, the new dataset counts <span style='color:grey'>`r nrow(df.MesoNeo)`</span> samples and <span style='color:grey'>`r length(unique(df.MesoNeo$culture))`</span> distinct cultures:

```{r pie_chart, echo=F, fig.align='center'}
nb.cols <- nrow(MNcultures)
mycolors <- colorRampPalette(brewer.pal(8, "Set1"))(nb.cols)
df.colors <- data.frame(Var1=unique(MNcultures$Var1),
                        color=mycolors)
data <- MNcultures %>% 
  arrange(desc(Var1)) %>%
  mutate(prop = Freq / sum(MNcultures$Freq) *100) %>%
  mutate(ypos = cumsum(prop)- 0.5*prop )

data <- merge(data, df.colors, by="Var1")
data$Var1 = factor(data$Var1, levels = data$Var1, ordered = T)
data$color = factor(data$color, levels = data$color, ordered = T)

ggplot(data, aes (x="", y = prop, fill = color)) + 
  geom_bar(width = 1, stat = "identity", color="white") +
  coord_polar("y")+
  geom_text(aes(label = Freq),
            position = position_stack(vjust = 0.5)) +
  scale_fill_identity(guide = "legend", name = "cultures", label = data$Var1) +
  blank_theme
```

The selected mtDNA sample of the dataset can be mapped. Colors of points are the same as previously. Inside their popup, the hg appears always colored in <span style="color: red;">red</span> but the culture is color is value-dependent   


```{r cldd, echo=F, fig.align='center', fig.cap="Cleaned [Ancient mtDNA database](https://amtdb.org/records/) using RShiny and Leaflet in R"}
df.MesoNeo.a <- df.MesoNeo
df.MesoNeo.a <- merge(df.MesoNeo.a, df.colors, by.x="culture", by.y="Var1", all.x=T)
# df.MesoNeo.a <- df.MesoNeo.a[!(df.MesoNeo.a$culture %in% unicums$Var1),]
desc <- paste0("<b>", df.MesoNeo.a$site, 
               "</b> <span style='color:red'>", df.MesoNeo.a$mt_hg, "</span> <br>",
               "culture: <span style='color:",df.MesoNeo.a$color,";'>", df.MesoNeo.a$culture, "</span> <i>",
               # "culture: <span style='color:orange'>", df.MesoNeo.a$culture, "</span> <i>",
               df.MesoNeo.a$epoch, "</i><br>")
df.MesoNeo.a$lbl <- paste0(desc, 'ref: <a href=',
                           shQuote(paste0(df.MesoNeo.a[,'reference_links'])),
                           "\ target=\"_blank\"",
                           ">", df.MesoNeo.a[,'reference_names'], "</a>")
leaflet(data = df.MesoNeo.a, width = "90%", height = "500px") %>%
  addTiles(group = 'OSM') %>%
  addCircleMarkers(layerId = ~identifier, 
                   lng = ~longitude,
                   lat = ~latitude,
                   weight = 1,
                   radius = 3,
                   popup = ~lbl,
                   color = ~color,
                   opacity = 0.7,
                   fillOpacity = 0.7) %>%
  addLegend("bottomleft", 
            colors = unique(df.MesoNeo.a$color),
            labels= unique(df.MesoNeo.a$culture),
            title = "cultures",
            opacity = 1)
```

A simple manner to model GC-coev is to create a co-occurences' graph of hg and cultures ('mt_hg' and 'culture' fields). The create graph will be a bipartite one (i.e, 2-mode graph, two classes of vertices). It could be created with the  [igraph](https://cran.r-project.org/web/packages/igraph/index.html) package. Graph drawing is a well-known heuristic, the default layout is a force-directed algorithm (e.g. Fruchterman-Reingold layout) allowing to bring  closer or move away vertices depending on the edges they share. It allows to manipulate easily the graphs. Here, graph vertices are either cultures or hg. There is a link between a culture and a hg  when an entity of the culture refers to a hg. 

```{r ddeodd, warning=F, echo=F, fig.width=10,fig.height=10}
par(mar = c(0, 0, 0, 0))
# df.MesoNeo$culture <- toupper(df.MesoNeo$culture)
edges <- df.MesoNeo.a[, c("culture", "mt_hg")]
# grph <- graph_from_data_frame(edges, directed = TRUE, vertices = NULL)
nds.culture <- as.character(data$Var1)
nds.culture.color <- as.character(data$color)
# nds.culture <- unique(df.MesoNeo$culture)
# nds.culture.color <- rep("gold", length(nds.culture))
nds.mt_hg<- unique(df.MesoNeo.a$mt_hg)
nds.mt_hg.color <- rep("grey", length(nds.mt_hg))
thm.ths.nds <- data.frame(name=c(nds.culture, nds.mt_hg),
                          color=c(nds.culture.color, nds.mt_hg.color),
                          stringsAsFactors = F)
g <- graph_from_data_frame(edges,
                           directed=F,
                           vertices= thm.ths.nds)
g <- simplify(g)
set.seed(123)
laout <- layout_with_fr(g)
plot(g, vertex.label.cex= 0.7, layout = laout,
     vertex.frame.color=NA,
     vertex.label.color = "black",
     vertex.label.family="Helvetica",
     vertex.size=9)
```   

A first reading of the graph shows that:

1. U5 hg characterize the HG group [@Bramanti09]
2. N1a hg characterize the EF group [@Brandt13]
3. H hg characterize the Late Neolithic groups [@Brandt13]

The dataset variability can be reduced with agglomeration techniques like the the [community detection algorithm](#com.det)

## Community detection {#com.det}

To to detect communities of each vertex, one can uses the 'fast greedy' algorithm  ([fastgreedy.community](https://www.rdocumentation.org/packages/igraph/versions/0.4.1/topics/fastgreedy.community) from the igraph package. This algorithm is a hierarchical ranking algorithm where initially each vertex belongs to a distinct community, and the communities are merged iteratively, so that each merge is locally optimal. The algorithm stops when it is no longer possible to increase the modularity, it will be thus gives a grouping as well as a dendrogram. This algorithm is close to the agglomeration of Ward used in hierarchical clustering (HC)

```{r frr, warning=F, echo=F, fig.width=10 ,fig.height=10}
# g <- simplify(g)
par(mar = c(0, 0, 0, 0))
g_community <- fastgreedy.community(g)
plot(g_community, layout = laout,
     g, vertex.label.cex= 0.7,
     vertex.frame.color=NA,
     vertex.label.color = "black",
     vertex.label.family="Helvetica",
     vertex.size=9)
```

A first reading of the clusters show a clear separation (i.e. edge distance) between three groups:

1. Hunter-Gatherers (Hunter-Gatherer)
2. Spain Early Neolithic (Iberia_EN)
3. Spain and France Middle/Late Neolithic (NE_Iberia_MLN and France_MLN) 

GC-coev needs to be refined. Maybe the cultural membership of mtDNA sample can be discussed upstream specifying the cultures they belong to, or downstream, choosing a different [community detection algorithm](#com.det), etc.

# mtDNA *and* radiocarbon dates

With their unique LabCode, the  dates associated with mtDNA samples can be easily be connected to existing [databases](https://neolithic.shinyapps.io/AbsoluteDating/#Databases). We will use the metadata .csv file [`r metadata`](mtdb_metadata.csv) to join the sample identifiers with the radiocarbon dates. These latter have a unique laboratory number ([LabCode](https://zoometh.github.io/C14/neonet/#mf.labcode)). This permit to associate the mtDNA sample with the radiocarbon date. 

The dataset needs to be cleaned keeping interesting fields and removing samples with typo errors (like longitude < - 90). For this training, the selected samples are those associated to radiocarbon dates: empty radiocarbon values and empty LabCode are removed. The radiocarbon field 'bp' is not strictly formatted, a text edition has to be done with regular expressions (*regex*) by splitting the values from 'bp' field (split on '±') to copy them in two: columns BP and SD. Finally, the analysis will still focus on Early Neolithic in West Europe (longitude < 7)

```{r ddeo, echo=T, message=F}
# selected.fields <- c("identifier", "mt_hg", "site", "culture", "epoch", "bp", "c14_lab_code", "year_from", "year_to", "longitude", "latitude", "reference_names", "reference_links")
selected.fields <- c("identifier", "site", "country", "culture", "epoch", "mt_hg", "c14_lab_code", "bp", "longitude", "latitude", "reference_names", "reference_links")
df <- read.csv(metadata, encoding = "UTF-8")
df <- df[, selected.fields]
df <- df %>% 
  # select(selected.fields) %>%
  filter(country == "Spain") %>%
  # filter(latitude > 35) %>%
  # filter(longitude > -6 & longitude < 5) %>%
  # filter(longitude >= -90 & longitude < 9) %>%
  # filter(longitude >= -90 & longitude <= 90) %>%
  # filter(latitude >= -90 & latitude <= 90) %>%
  filter(mt_hg != "" & !is.na(mt_hg)) %>%
  filter(epoch == 'Neolithic' | epoch == 'Mesolithic') %>%
  filter(bp != "" & !is.na(bp))  %>%
  filter(c14_lab_code != "" & !is.na(c14_lab_code))  %>%
  filter(str_detect(bp, "±"))  %>%
  filter(!str_detect(bp, "BP|and")) # %>%
# filter(year_from >= -6500 & year_from <= -4500)
bps <- unlist(sapply(df$bp,
                     function(x) strsplit(x, "±")),
              use.names = FALSE)
mat <- matrix(bps, ncol=2, byrow=TRUE)
df$c14bp <- as.integer(mat[, 1])
df$c14sd <- as.integer(mat[, 2])
df$bp <- NULL
```

Now the new dataset has <span style="color: grey;">`r nrow(df)`</span> samples and <span style="color: grey;">`r ncol(df)`</span> columns:

```{r c14, echo=F, fig.align='center'}
kable(df[sample(nrow(df), 6), ],"html",
      row.names = F,
      caption = "aDNA metadata sample") %>%
  kable_styling(full_width = FALSE, position = "center", font_size=11)
```

The dataset can be spatialized

```{r map_ok, echo=F, fig.align='center', fig.cap="Cleaned [Ancient mtDNA database](https://amtdb.org/records/) using RShiny and Leaflet in R"}
MNcultures <- as.data.frame(table(df$culture),
                            stringsAsFactors = F)
nb.cols <- nrow(MNcultures)
mycolors <- colorRampPalette(brewer.pal(8, "Set1"))(nb.cols)
df.colors <- data.frame(Var1=unique(MNcultures$Var1),
                        color=mycolors)
df.MesoNeo.b <- merge(df, df.colors, by.x="culture", by.y="Var1", all.x=T)
desc <- paste0("<b>", df.MesoNeo.b$site, "</b> <span style='color:red'>",
               df.MesoNeo.b$mt_hg, "</span> <br>",
               "culture: <span style='color:",df.MesoNeo.b$color,";'>", 
               df.MesoNeo.b$culture, "</span> <i>", df.MesoNeo.b$epoch, "</i><br>", 
               "c14: ", df.MesoNeo.b$c14bp, " ± ", df.MesoNeo.b$c14sd, " BP ", 
               "[", df.MesoNeo.b$c14_lab_code, "]")
df.MesoNeo.b$lbl <- paste0(desc, '<br> ref: <a href=', 
                           shQuote(paste0(df.MesoNeo.b[,'reference_links'])),"\ target=\"_blank\"",
                           ">", df.MesoNeo.b[,'reference_names'], "</a>")
leaflet(data = df.MesoNeo.b, width = "60%", height = "400px") %>%
  # setView(lng = -111.846061, lat = 36.115847, zoom = 12) %>%
  addTiles(group = 'OSM') %>%
  addCircleMarkers(layerId = ~identifier, 
                   lng = ~longitude,
                   lat = ~latitude,
                   weight = 1,
                   radius = 3,
                   popup = ~lbl,
                   color = ~color,
                   opacity = 0.7,
                   fillOpacity = 0.7) %>%
  addLegend("bottomleft", 
            colors = unique(df.MesoNeo.b$color),
            labels= unique(df.MesoNeo.b$culture),
            title = "cultures",
            opacity = 1)
```

R offers packages to calibrate dates like [Bchron](https://cran.r-project.org/web/packages/Bchron/index.html) or [rcarbon](https://cran.r-project.org/web/packages/rcarbon/index.html)

```{r calib.all, echo=T, fig.align='center', fig.width=12 ,fig.height=7, fig.cap="Calibration of multiple radiocarbon date with [rcarbon](https://cran.r-project.org/web/packages/rcarbon/index.html)"}
c14dates <- calibrate(x=df.MesoNeo.b$c14bp,
                      errors=df.MesoNeo.b$c14sd,
                      calCurves='intcal20',
                      ids=df.MesoNeo.b$c14_lab_code)
multiplot(c14dates,decreasing=TRUE,rescale=TRUE,HPD=TRUE)
```

### Comtemporaneous mtDNA samples

If one wants to study contemporaneous mtDNA samples, a statistical test has be performed. At first, we can discard the date with the LabCode MAMS-14585 which it clearly isolated, and replot the rest of the dates. We add the identifiers of mtDNA samples on each summed probability distributions

```{r calib.sel, echo=T, fig.align='center', fig.width=12 ,fig.height=7, fig.cap="Calibration of multiple radiocarbon date with [rcarbon](https://cran.r-project.org/web/packages/rcarbon/index.html)"}
df.MesoNeo.c <- df.MesoNeo.b[df.MesoNeo.b$c14_lab_code != 'MAMS-14585', ]
idfs <- paste0(df.MesoNeo.c$c14_lab_code,"\n",df.MesoNeo.c$identifier)
c14dates <- calibrate(x=df.MesoNeo.c$c14bp,
                      errors=df.MesoNeo.c$c14sd,
                      calCurves='intcal20',
                      ids=idfs)
multiplot(c14dates,decreasing=TRUE,rescale=TRUE,HPD=TRUE)
```

Then we will perform a Mann-Whitney test (function [wilcox.test](https://www.rdocumentation.org/packages/stats/versions/3.6.2/topics/wilcox.test)) on the ageGrid (results of the function [BchronCalibrate](https://rdrr.io/cran/Bchron/man/BchronCalibrate.html)) for each radiocarbon date pairwise. When for a pairwise the p-value is superior to 0.05 (95%), the two dates are considered contemporaneous 

```{r wilcox, echo=T, fig.height=7, fig.width=7, fig.align='center', fig.cap="Mann-Whitney test results on grid ages extents of all dates pairwise"}
l <- list()
for(i in 1:nrow(df.MesoNeo.c)){
  ids <- df.MesoNeo.c[i, "identifier"]
  a.date <- BchronCalibrate(ages=df.MesoNeo.c[i, "c14bp"],
                            ageSds=df.MesoNeo.c[i, "c14sd"],
                            calCurves='intcal20',
                            ids=df.MesoNeo.c[i, "identifier"])
  lst <- list(eval(parse(text=paste0("a.date$",ids,"$ageGrid"))))
  names(lst) <- ids
  l[length(l)+1] <- lst
}
n.obs <- sapply(l, length)
seq.max <- seq_len(max(n.obs))
mat <- t(sapply(l, "[", i = seq.max))
rownames(mat) <- df.MesoNeo.c[, "identifier"]
mat.t <- t(mat)
n.col <- 1:length(rownames(mat))
lcompar <- t(combn(n.col, 2))
df.test <- data.frame(a = character(0),
                      b = character(0),
                      pval = integer(0))
for(i in 1:nrow(lcompar)){
  p1 <- lcompar[i, 1]
  p2 <- lcompar[i, 2]
  a <- colnames(mat.t)[p1]
  b <- colnames(mat.t)[p2]
  pval <- round(wilcox.test(mat.t[, p1], mat.t[, p2])$p.value, 3)
  df.test <- rbind(df.test, c(a, b, pval))
  df.test <- rbind(df.test, c(b, a, pval))
}
colnames(df.test) <- c("Var1", "Var2", "value")
df.test$value <- as.numeric(df.test$value)
ggplot(data = df.test, aes(x=Var1, y=Var2, fill=value)) +
  geom_tile(na.rm = T) +
  scale_fill_gradient2(low = "yellow", high = "red", space = "Lab",
                       na.value = 'white')+
  geom_text(aes(Var1, Var2, label = value),
            color = "black", size = 2.5)+
  theme_minimal()+
  theme_bw()+
  theme(axis.title = element_blank())+
  theme(axis.text=element_text(size=8))+
  theme(axis.ticks = element_blank())+
  theme(axis.ticks.length=unit(.1, "cm"))+
  theme(legend.position = "none")
```

Only two mtDNA/radiocarbon dates have homogenized grid ages extents and can be considerated has contemporaneous, I0413 and I0410

```{r c14.contemp, echo=F, fig.align='center'}
selected.fields <- c("site","country","culture","identifier", "mt_hg", "c14_lab_code", "c14bp", "c14sd", "reference_names", "reference_links")
contemp.df <- df.MesoNeo.c[df.MesoNeo.c$identifier %in% c('I0413', 'I0410'), selected.fields]
kable(contemp.df,"html",
      row.names = F,
      caption = "Contemporaneus mtDNA samples") %>%
  kable_styling(full_width = FALSE, position = "center", font_size=11)
```

# References
